OpenSim/Simulation/StatesTrajectory.h

```diff
@@ -0,0 +1,641 @@
+ * appending states to it, but users cannot modify the individual states that
+ * are already in the trajectory.
+ *
+ * This class was introduced in OpenSim version 4.0, and enables scripting
+ * (Python/MATLAB) and C++ users to postprocess their results much more
+ * flexibly than with an Analysis.
+ *
+ * ### Guarantees
+ * This class is designed to ensure the following:
+ * - The states are ordered nondecreasing in time (adjacent states *can* have
+ *   the same time).
+ * - All states in the trajectory are consistent with each other (see
+ *   isConsistent()).
+ *
+ * @note These gaurantees apply when using this class through C++, Java,
+ * or the %OpenSim GUI, but **not** through Python or MATLAB. This is because
+ * Python and MATLAB do not enforce constness and thus allow modifying the
+ * trajectory.
+ *
+ * ### Using with a %Model 
+ * A StatesTrajectory is not very useful on its own, since neither the
+ * trajectory nor the contained states know of the names that the Component%s
+ * give to the state variables. You probably want to use the trajectory with an
+ * OpenSim::Model, through which the state variables have a meaning (e.g.,
+ * `model.getStateVariableValue(states[0], "soleus_r/activation")`).
+ *
+ * SimTK::State%s have a tight association with a specific OpenSim::Model
+ * (actually, with the SimTK::System within an OpenSim::Model). However,
+ * neither the StatesTrajectory nor the OSTATES file knows the model to which
+ * it corresponds. So, for example, you could use a single StatesTrajectory
+ * with a generic gait2392 model as well as with a scaled (subject-specific)
+ * gait2392 model. This flexibility may be beneficial in some scenarios, but
+ * also allows one to accidentally use the wrong model with a given states
+ * trajectory, potentially leading to silent errors that could compromise a
+ * scientific study.
+ *
+ * To increase your confidence that a StatesTrajectory matches a given Model,
+ * you can perform some weak checks with isCompatibleWith().
+ *
+ * TODO acceleration-level calculations?
+ *
+ * ### File format
+ * StatesTrajectory files use the file extension `.OSTATES`, with the XML
+ * format. Therefore, you could theoretically edit a StatesTrajectory file in
+ * a typical text editing program, or in Python/MATLAB using XML libraries
+ * (such a process is is likely to be painful; consider using the
+ * createFromStatesStorage() utility instead).
+ *
+ * A SimTK::State object contains many different types of data, but only some
+ * are saved into the OSTATES file:
+ * 
+ * type of data                 | saved in OSTATES?
+ * ---------------------------- | -------------
+ * (continuous) state variables | yes
+ * discrete state variables     | yes
+ * modeling options             | yes
+ * cache variables              | no
+ *
+ * The cache variables (e.g., total system mass, control signals) are not saved
+ * to the OSTATES file because they can be regenerated from the other data and
```

Comment:
could refer to the definition of a cache variable as being a variable that can be computed/realized from state values and is conceptually separate from the state (although it is tucked in the State because they are intimately related - cache on state).